controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -72,26 +78,48 @@ func (c *EtcdClientGenerator) forFirstAvailableNode(ctx context.Context, nodeNam
 
 // forLeader takes a list of nodes and returns a client to the leader node.
 func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
-	var errs []error
+	// This is an additional safeguard for avoiding this func to return nil, nil.
+	if len(nodeNames) == 0 {
+		return nil, errors.New("invalid argument: forLeader can't be called with an empty list of nodes")
+	}
 
+	// Loop through the existing control plane nodes.
+	var errs = make([]error, 0)
 	for _, nodeName := range nodeNames {
+		// Get a temporary client to the etcd instance hosted on the node.
 		client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
 		defer client.Close()
+
+		// Get the list of members.
 		members, err := client.Members(ctx)
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
+
+		// Loop through the list of etcd members reported checking for the current leader.
 		for _, member := range members {
-			if member.Name == nodeName && member.ID == client.LeaderID {
-				return c.forFirstAvailableNode(ctx, []string{nodeName})
+			if member.ID == client.LeaderID {
```

Comment:
nit: maybe flatten the logic a bit. One for loop to find the leader member and if we found one another to call `forFirstAvailableNode` with the right nodeName (or actually it just verifies that the leader member.Name is in nodeNames, maybe we should make that explicit)

Reply:
not sure I fully understand what you are suggesting...

Reply:
Replacing l.104-114 with:
```
		// Loop through the list of etcd members reported checking for the current leader.
		var leaderMember *etcd.Member
		for _, member := range members {
			if member.ID == client.LeaderID {
				leaderMember = member
			}
		}

		// If we found the leader and it is one of the nodes,
		// get a connection to the etcd leader via the node hosting it.
		if leaderMember != nil && nodes.Has(leaderMember.Name) {
			return c.forFirstAvailableNode(ctx, []string{nodeName})
		}
```

Somewhere above:
```go
	nodes := sets.NewString()
	for _, n := range nodeNames {
		nodes.Insert(n)
	}
```

Just feels to me, like it's easier to understand the underlying logic that way

Reply:
thanks for clarifying!
done!